features/nfc/doc/nfc_design.md

```diff
@@ -0,0 +1,424 @@
+
+In this case buffers are copied to account for the NULL-terminator character that is not present in the underlying byte buffer.
+
+## HAL APIs
+
+### NFC EEPROM API
+
+The one HAL API that will have to be implemented by vendors to make use of the `NFCEEPROM` class are the following virtual methods.
+
+From the upper layer's point of view, the EEPROM is a byte array that can be read from/written to. Long operations (reads, writes, erasures) must happen asynchronously. Booleans indicate whether a particular operation was succesful. Encoding is handled by the upper layer.
+
+Address 0 means the start of the NDEF buffer (not necessarily at address 0 in the EEPROM). 
+
+When a buffer is passed to the backend, the reference remains valid till the corresponding event is called.
+
+The `backend_set_size()` command is called to change the size of the buffer (within the limits set by `backend_get_max_size()`). Inversely that buffer size can be read by `backend_get_size()`.
+
+```cpp
+void backend_reset()
+size_t backend_get_max_size() const
+void backend_read_bytes(uint32_t address, size_t count)
+void backend_write_bytes(uint32_t address, const uint8_t* bytes, size_t count)
+void backend_set_size(size_t count)
+void backend_get_size()
+void backend_erase_bytes(uint32_t address, size_t size) 
+```
+
+The following events must be called to signal completion of long operations:
+```cpp
+void on_backend_has_read_bytes(bool success, const uint8_t* bytes)
+void on_backend_has_written_bytes(bool success)
+void on_backend_has_set_size(bool success)
+void on_backend_has_gotten_size(bool success, size_t size)
+void on_backend_has_erased_bytes(bool success)
+```
+
+### NCI Driver APIs
+
+This API will be defined in phase 2.
+
+# Testing strategy
+## NFC Forum Compliance
+
+A dongle driven by [PyNFC](https://nfcpy.readthedocs.io/en/latest/index.html) will be used to run GreenTea-based tests to ensure that the implementation behaves correctly for a range of system tests.
+
+Unit tests will cover all internal logic and NFC endpoints can be mocked/emulated where possible.
+
+In the future we could run NFC Forum test suites using approved testing equipment.
+
+## Interoperability
+
+Interoperability is important with a technology such as NFC. Therefore our testing rig will include a selection of smartphones and NFC tags that can be connected using analog switches to the relevant NFC-enabled platform running mbed OS.
+
+![interop_test_rig]
+
+## HAL testing
+
+GreenTea tests will be provided to partners to ensure compliance with the NFC EEPROM backend API.
+
+# Dependencies
+* Event Queue
+
+There are currently at least four event queues (Plaftorm, BLE, USB, IP) in mbed OS and NFC will also require an event queing mechanism. We should aim at reusing one of these existing queues with the long term goal of unifying these code bases.
```

Comment:
This can be solved by not instantiating the event queue inside the NFC stack but rather let the application provide it. 
Dependency injection delegate the layout of the program to the application writer. Ble already does that for this reason.